arrows/ocv/refine_detections_draw.cxx

```diff
@@ -0,0 +1,183 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * \file
+ * \brief Implementation of OCV draw tracks algorithm
+ */
+
+#include "refine_detections_draw.h"
+
+#include <fstream>
+#include <sstream>
+#include <iostream>
+#include <iomanip>
+#include <algorithm>
+#include <deque>
+
+#include <vital/vital_config.h>
+#include <vital/vital_foreach.h>
+#include <vital/logger/logger.h>
+#include <vital/exceptions/io.h>
+
+#include <kwiversys/SystemTools.hxx>
+
+#include <arrows/ocv/image_container.h>
+
+#include <opencv2/core/core.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+#include <opencv2/highgui/highgui.hpp>
+
+using namespace kwiver::vital;
+
+namespace kwiver {
+namespace arrows {
+namespace ocv {
+
+typedef kwiversys::SystemTools ST;
+
+/// Private implementation class
+class refine_detections_draw::priv
+{
+public:
+
+  /// Constructor
+  priv()
+  : pattern( "feature_tracks_%05d.png" ),
+    id( 0 )
+  {
+  }
+
+  /// Destructor
+  ~priv()
+  {
+  }
+
+  /// Parameters
+  std::string pattern;
+
+  /// Variables
+  unsigned id;
+};
+
+
+/// Constructor
+refine_detections_draw
+::refine_detections_draw()
+: d_( new priv() )
+{
+}
+
+
+/// Destructor
+refine_detections_draw
+::~refine_detections_draw()
+{
+}
+
+
+/// Get this algorithm's \link vital::config_block configuration block \endlink
+vital::config_block_sptr
+refine_detections_draw
+::get_configuration() const
+{
+  vital::config_block_sptr config = vital::algo::refine_detections::get_configuration();
+
+  config->set_value( "pattern", d_->pattern,
+                     "The output pattern for writing images to disk." );
+
+  return config;
+}
+
+
+/// Set this algorithm's properties via a config block
+void
+refine_detections_draw
+::set_configuration( vital::config_block_sptr in_config )
+{
+  vital::config_block_sptr config = this->get_configuration();
+  config->merge_config( in_config );
+
+  d_->pattern = config->get_value<std::string>( "pattern" );
+}
+
+/// Check that the algorithm's currently configuration is valid
+bool
+refine_detections_draw
+::check_configuration(vital::config_block_sptr config) const
+{
+  return true;
+}
+
+/// Output images with tracked features drawn on them
+vital::detected_object_set_sptr
+refine_detections_draw
+::refine( vital::image_container_sptr image_data,
+          vital::detected_object_set_sptr detections ) const
+{
+  cv::Mat img = ocv::image_container::vital_to_ocv( image_data->get_image() );
+
+  VITAL_FOREACH( auto det, detections->select() )
+  {
+    // Generate output filename
+    std::string ofn;
+    size_t max_len = d_->pattern.size() + 4096;
+    ofn.resize( max_len );
+    int num_bytes = snprintf( &ofn[0], max_len, d_->pattern.c_str(), d_->id++ );
+
+    if( num_bytes < 0 )
+    {
+      LOG_WARN( logger(), "Could not format output file name: \"" <<
+        d_->pattern << "\". Disabling writing to disk." );
```

Comment:
Nothing good is going to happen if the file name string is empty. Should this abort or continue?